plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,242 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Răzvan Vîrtan <virtanrazvan@gmail.com>
+ *          Justin He     <justin.he@arm.com>
+ *
+ * Copyright (c) 2021, Arm Ltd. All rights reserved.
+ * Copyright (c) 2021, University Politehnica of Bucharest. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <kvm/config.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <kvm-arm/smp_defs.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static uint64_t bsp_cpu_id;
+static unsigned int smp_numcores;
+void _lcpu_start(struct arm64_cpu *cpu);
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, cell;
+	unsigned int cpu_idx;
+	int subnode;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
```

Comment:
TODO: Returning from this function with return -1 leaves some
important global state uninitialized!
Better: set bsp_cpu_id and smp_numcores=1 also in case of error.
Then lcpu_start can be called without harm as other CPUs are
simply out of range.
BUT: Can it be that bsp_cpu_id won't be 0? In that case we have a
more complex scenario here, as some CPU x with x < bsp_cpu_id may
had problems in discovery. That means smp_numcores <= bsp_cpu_id.
What is a valid value for smp_numcores then? bsp_cpu_id must
always be < smp_numcores!

ANSWER: bsp_cpu_id represents a "physical id" (an id read in `ukplat_lcpu_id()`
by using the affinity bits in MPIDR). smp_numcores, on the other side, 
represents the maximum logic id of an available cpu (the maximum index in the cpus
array where we can find an available cpu). Since the condition we check
in `ukplat_lcpu_start` for each AP is `logic_id < smp_numcores`, I don't think
we risk to use an unavailable cpu, even if it's physical id is <= than the
boot cpu id.

Reply:
> BUT: Can it be that bsp_cpu_id won't be 0? 
Yes, cpu ID is implementation specific and you might have a bootloader which doesn't load on CORE 0 (can be unusual, but I've seen this in real world), we shouldn't make any assumption about its value. I think you should just try to initialize all CPUs described in DTS, AFAIK, by PSCI spec, if PSCI is supported, CPU_ON must be present... or at least should return an error if for some reason (for instance, wrong ID in DTS) that CPU is not available...